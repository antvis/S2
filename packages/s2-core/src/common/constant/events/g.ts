export enum GEventType {
  RIGHT_MOUSE_UP = 'rightup',
}
