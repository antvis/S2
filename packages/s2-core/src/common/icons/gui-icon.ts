/**
 * @Description: 请严格要求 svg 的 viewBox，若设计产出的 svg 不是此规格，请叫其修改为 '0 0 1024 1024'
 */
import { Group, Shape, ShapeAttrs } from '@antv/g-canvas';
import { omit, clone } from 'lodash';
import { getIcon } from './factory';

const STYLE_PLACEHOLDER = '<svg';
// Image 缓存
const ImageCache: Record<string, HTMLImageElement> = {};

export interface GuiIconCfg extends ShapeAttrs {
  readonly name: string;
}

/**
 * 使用 iconfont 上的 svg 来创建 Icon
 */
export class GuiIcon extends Group {
  // icon 对应的 GImage 对象
  private image: Shape.Image;

  constructor(cfg: GuiIconCfg) {
    super(cfg);
    this.render();
  }

  // 获取 Image 实例，使用缓存，以避免滚动时因重复的 new Image() 耗时导致的闪烁问题
  /* 异步获取 image 实例 */
  private getImage(
    name: string,
    cacheKey: string,
    fill?: string,
  ): Promise<HTMLImageElement> {
    return new Promise<HTMLImageElement>((resolve, reject): void => {
      const img = new Image();
      // 成功
      img.onload = () => {
        ImageCache[cacheKey] = img;
        resolve(img);
      };
      // 失败
      img.onerror = (e) => {
        reject(e);
      };
      let svg = getIcon(name);

      // 兼容三种情况
      // 1、base 64
      // 2、svg本地文件（兼容老方式，可以改颜色）
      // 3、线上支持的图片地址
      if (
        svg &&
        (svg.includes('data:image/svg+xml') || this.hasSupportSuffix(svg))
      ) {
        // 传入 base64 字符串
        // 或者 online 链接
        img.src = svg;
      } else if (svg) {
        // 传入 svg 字符串（支持颜色fill）
        if (fill) {
          // 如果有fill，移除原来的 fill
          // 这里有一个潜在的问题，不同的svg里面的schema不尽相同，导致这个正则考虑不全
          // 1、fill='' 2、fill 3、fill-***(不需要处理)
          // eslint-disable-next-line no-useless-escape
          svg = svg.replace(/fill=[\'\"]#?\w+[\'\"]/g, ''); // 移除 fill="red|#fff"
          svg = svg.replace(/fill>/g, '>'); // fill> 替换为 >
        }
        // https://www.chromestatus.com/features/5656049583390720
        // # 井号不能当做svg的body，这个bug在chrome72已经修复.
        svg = svg
          .replace(STYLE_PLACEHOLDER, `${STYLE_PLACEHOLDER} fill="${fill}"`)
          .replace(/#/g, '%23');
        img.src = `data:image/svg+xml;utf-8,${svg}`;
      }
    });
  }

  hasSupportSuffix = (image: string) => {
    return ['.png', '.jpg', '.gif', '.svg'].some((suffix) =>
      image?.endsWith(suffix),
    );
  };

  private render() {
    const { name, fill } = this.cfg;
    const attrs = clone(this.cfg);
    const image = new Shape.Image({
      attrs: omit(attrs, 'fill'),
    });

    const cacheKey = `${name}-${fill}`;
    const img = ImageCache[cacheKey];
    if (img) {
      // already in cache
      image.attr('img', img);
      this.addShape('image', image);
    } else {
      this.getImage(name, cacheKey, fill)
        .then((value: HTMLImageElement) => {
          image.attr('img', value);
          this.addShape('image', image);
        })
        .catch((err: Event) => {
          // eslint-disable-next-line no-console
          console.warn(`GuiIcon ${name} load error`, err);
        });
    }
    this.image = image;
  }
}
