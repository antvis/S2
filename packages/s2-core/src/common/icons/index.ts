export { getIcon, registerIcon } from './factory';
export { GuiIcon } from './gui-icon';
