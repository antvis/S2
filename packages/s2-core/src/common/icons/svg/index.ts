// 不本地直接引用svg，避免下游无依赖
export {
  ArrowDown,
  ArrowUp,
  CellDown,
  CellUp,
  GlobalAsc,
  GlobalDesc,
  GroupAsc,
  GroupDesc,
  GroupNone,
  Minus,
  Plus,
  SortDown,
  SortUp,
  SortDownSelected,
  SortUpSelected,
  InfoCircle,
  ExpandColIcon,
  Trend,
  DrillDownIcon,
  EyeOutlined,
} from './svgs';
