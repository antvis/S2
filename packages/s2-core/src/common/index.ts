export * from './constant';
export * from './debug';
export * from './i18n';
export * from './icons';
export * from './interface';
export * from './store';
