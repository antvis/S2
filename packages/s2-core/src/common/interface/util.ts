export type DeepRequired<T extends Record<string, any>> = {
  [K in keyof T]-?: NonNullable<T[K]> extends Record<string, any>
    ? DeepRequired<NonNullable<T[K]>>
    : NonNullable<T[K]>;
};
