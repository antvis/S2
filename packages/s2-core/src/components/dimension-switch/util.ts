import { DimensionType, DimensionItem } from './dimension';

export interface OperatedType {
  [type: string]: string[];
}

export const getDimensionsByPredicate = (
  data: DimensionType[],
  predicate: (value: DimensionItem) => boolean,
) => {
  return data.reduce((obj, t) => {
    obj[t.type] = t.items.filter(predicate).map((i) => i.id);
    return obj;
  }, {} as OperatedType);
};
