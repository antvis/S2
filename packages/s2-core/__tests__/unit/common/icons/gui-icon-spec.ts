import { Group } from '@antv/g';
import { registerIcon } from '../../../../src/common/icons';
import { sleep, createPivotSheet } from '../../../util/helpers';
import { CustomImage } from '@/engine/CustomImage';
import { GuiIcon } from '@/common/icons/gui-icon';
import { ArrowDown } from '@/common/icons/svg/svgs';

describe('GuiIcon Tests', () => {
  test('should get gui icon static type', () => {
    expect(GuiIcon.type).toEqual('__GUI_ICON__');
  });

  test.each([
    // 内置
    'SortUp',
    // base64/本地文件
    ArrowDown,
    // 在线链接 (无后缀)
    'https://mdn.alipayobjects.com/huamei_qa8qxu/afts/img/A*2VvTSZmI4vYAAAAAAAAAAAAADmJ7AQ/original',
    // 在线链接 (静态)
    'https://mdn.alipayobjects.com/huamei_qa8qxu/afts/img/A*P-jqT4U7YrcAAAAAAAAAAAAADmJ7AQ/original.jpg',
    // 在线链接 (动态)
    'https://mdn.alipayobjects.com/huamei_qa8qxu/afts/img/A*f6e6S4OUSdMAAAAAAAAAAAAADmJ7AQ/original.gif',
    // 在线链接 (webp)
    'https://mdn.alipayobjects.com/huamei_qa8qxu/afts/img/A*5nsESLuvc_EAAAAAAAAAAAAADmJ7AQ/fmt.webp',
  ])('should render correctly icon with %s', (src) => {
    const errSpy = jest
      .spyOn(console, 'error')
      .mockImplementationOnce(() => {});

    registerIcon('test', src);

    const icon = new GuiIcon({
      name: 'test',
      x: 0,
      y: 0,
      width: 20,
      height: 20,
    });

    expect(icon.name).toEqual('test');
    expect(icon.iconImageShape).toBeInstanceOf(CustomImage);
    expect(icon).toBeInstanceOf(Group);
    expect(errSpy).not.toHaveBeenCalled();
  });

  test('should not render icon with invalid online url', async () => {
    registerIcon('test', 'https://www.test.svg');

    const errSpy = jest
      .spyOn(console, 'error')
      .mockImplementationOnce(() => {});

    // eslint-disable-next-line no-new
    new GuiIcon({
      name: 'test',
      x: 0,
      y: 0,
      width: 20,
      height: 20,
    });

    await sleep(300);

    expect(errSpy).toHaveBeenCalled();
  });

  test('should get is online link result', () => {
    const icon = new GuiIcon({
      name: 'test',
      x: 0,
      y: 0,
      width: 20,
      height: 20,
    });

    expect(icon.isOnlineLink('https://www.test.png')).toBeTruthy();
    expect(icon.isOnlineLink('https://www.test/test')).toBeTruthy();
    expect(icon.isOnlineLink('http://www.test.png')).toBeTruthy();
    expect(icon.isOnlineLink('//www.test.png')).toBeTruthy();
    expect(icon.isOnlineLink('https//www.test.png')).toBeFalsy();
    expect(icon.isOnlineLink('https//www.test.png')).toBeFalsy();
    expect(icon.isOnlineLink('://www.test.png')).toBeFalsy();
    expect(icon.isOnlineLink('')).toBeFalsy();
  });

  test('should be able to update image with setImageAttr function.', () => {
    registerIcon('test', 'SortUp');

    const icon = new GuiIcon({
      name: 'test',
      x: 0,
      y: 0,
      width: 20,
      height: 20,
    });

    const spy = jest.spyOn(icon, 'getImage');
    const oldVal = icon.iconImageShape.style.img;

    expect(oldVal).toBeDefined();
    icon.setImageAttrs({ fill: 'red' });
    expect(spy).toHaveBeenCalled();
  });

  // https://github.com/antvis/S2/issues/2513
  test('should support cross origin for online url', () => {
    const s2 = createPivotSheet({
      width: 200,
      height: 200,
      customSVGIcons: [
        {
          name: 'Filter',
          svg: 'https://gw.alipayobjects.com/zos/antfincdn/gu1Fsz3fw0/filter%26sort_filter.svg',
        },
      ],
      headerActionIcons: [
        {
          icons: ['Filter'],
          belongsCell: 'colCell',
          defaultHide: false,
        },
        {
          icons: ['Filter'],
          belongsCell: 'rowCell',
          defaultHide: false,
        },
        {
          icons: ['Filter'],
          belongsCell: 'cornerCell',
          defaultHide: false,
        },
      ],
    });

    async function render() {
      await s2.render();
      s2.getCanvasElement().toDataURL();
    }

    expect(render).not.toThrow();
  });
});
