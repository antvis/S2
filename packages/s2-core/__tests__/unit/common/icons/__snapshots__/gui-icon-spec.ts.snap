// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GuiIcon Tests should render correctly icon with 
<svg width="10px" height="10px" viewBox="0 0 10 10" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
<g id="页面-1" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd" font-family="Roboto, PingFangSC, BlinkMacSystemFont, Microsoft YaHei, Arial, sans-serif" font-size="12" font-weight="bold">
    <text id="↓" fill="#29A294">
        <tspan x="-0.225097656" y="8.5">↓</tspan>
    </text>
</g>
</svg> 1`] = `
GuiIcon {
  "_events": Object {},
  "attrs": Object {
    "matrix": null,
    "opacity": 1,
  },
  "cfg": Object {
    "animable": true,
    "animating": false,
    "capture": true,
    "children": Array [],
    "height": 20,
    "name": "test",
    "visible": true,
    "width": 20,
    "x": 0,
    "y": 0,
    "zIndex": 0,
  },
  "destroyed": false,
  "iconImageShape": ImageShape1 {
    "_events": Object {},
    "attrs": Object {
      "animable": true,
      "animating": false,
      "capture": true,
      "children": Array [],
      "fillOpacity": 1,
      "height": 20,
      "lineAppendWidth": 0,
      "lineWidth": 1,
      "matrix": null,
      "name": "test",
      "opacity": 1,
      "strokeOpacity": 1,
      "type": "__GUI_ICON__",
      "visible": true,
      "width": 20,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
    "cfg": Object {
      "animable": true,
      "animating": false,
      "attrs": Object {
        "animable": true,
        "animating": false,
        "capture": true,
        "children": Array [],
        "height": 20,
        "name": "test",
        "type": "__GUI_ICON__",
        "visible": true,
        "width": 20,
        "x": 0,
        "y": 0,
        "zIndex": 0,
      },
      "capture": true,
      "visible": true,
      "zIndex": 0,
    },
    "destroyed": false,
  },
  "isOnlineLink": [Function],
}
`;

exports[`GuiIcon Tests should render correctly icon with SortUp 1`] = `
GuiIcon {
  "_events": Object {},
  "attrs": Object {
    "matrix": null,
    "opacity": 1,
  },
  "cfg": Object {
    "animable": true,
    "animating": false,
    "capture": true,
    "children": Array [],
    "height": 20,
    "name": "test",
    "visible": true,
    "width": 20,
    "x": 0,
    "y": 0,
    "zIndex": 0,
  },
  "destroyed": false,
  "iconImageShape": ImageShape1 {
    "_events": Object {},
    "attrs": Object {
      "animable": true,
      "animating": false,
      "capture": true,
      "children": Array [],
      "fillOpacity": 1,
      "height": 20,
      "lineAppendWidth": 0,
      "lineWidth": 1,
      "matrix": null,
      "name": "test",
      "opacity": 1,
      "strokeOpacity": 1,
      "type": "__GUI_ICON__",
      "visible": true,
      "width": 20,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
    "cfg": Object {
      "animable": true,
      "animating": false,
      "attrs": Object {
        "animable": true,
        "animating": false,
        "capture": true,
        "children": Array [],
        "height": 20,
        "name": "test",
        "type": "__GUI_ICON__",
        "visible": true,
        "width": 20,
        "x": 0,
        "y": 0,
        "zIndex": 0,
      },
      "capture": true,
      "visible": true,
      "zIndex": 0,
    },
    "destroyed": false,
  },
  "isOnlineLink": [Function],
}
`;

exports[`GuiIcon Tests should render correctly icon with https://mdn.alipayobjects.com/huamei_qa8qxu/afts/img/A*2VvTSZmI4vYAAAAAAAAAAAAADmJ7AQ/original 1`] = `
GuiIcon {
  "_events": Object {},
  "attrs": Object {
    "matrix": null,
    "opacity": 1,
  },
  "cfg": Object {
    "animable": true,
    "animating": false,
    "capture": true,
    "children": Array [],
    "height": 20,
    "name": "test",
    "visible": true,
    "width": 20,
    "x": 0,
    "y": 0,
    "zIndex": 0,
  },
  "destroyed": false,
  "iconImageShape": ImageShape1 {
    "_events": Object {},
    "attrs": Object {
      "animable": true,
      "animating": false,
      "capture": true,
      "children": Array [],
      "fillOpacity": 1,
      "height": 20,
      "lineAppendWidth": 0,
      "lineWidth": 1,
      "matrix": null,
      "name": "test",
      "opacity": 1,
      "strokeOpacity": 1,
      "type": "__GUI_ICON__",
      "visible": true,
      "width": 20,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
    "cfg": Object {
      "animable": true,
      "animating": false,
      "attrs": Object {
        "animable": true,
        "animating": false,
        "capture": true,
        "children": Array [],
        "height": 20,
        "name": "test",
        "type": "__GUI_ICON__",
        "visible": true,
        "width": 20,
        "x": 0,
        "y": 0,
        "zIndex": 0,
      },
      "capture": true,
      "visible": true,
      "zIndex": 0,
    },
    "destroyed": false,
  },
  "isOnlineLink": [Function],
}
`;

exports[`GuiIcon Tests should render correctly icon with https://mdn.alipayobjects.com/huamei_qa8qxu/afts/img/A*5nsESLuvc_EAAAAAAAAAAAAADmJ7AQ/fmt.webp 1`] = `
GuiIcon {
  "_events": Object {},
  "attrs": Object {
    "matrix": null,
    "opacity": 1,
  },
  "cfg": Object {
    "animable": true,
    "animating": false,
    "capture": true,
    "children": Array [],
    "height": 20,
    "name": "test",
    "visible": true,
    "width": 20,
    "x": 0,
    "y": 0,
    "zIndex": 0,
  },
  "destroyed": false,
  "iconImageShape": ImageShape1 {
    "_events": Object {},
    "attrs": Object {
      "animable": true,
      "animating": false,
      "capture": true,
      "children": Array [],
      "fillOpacity": 1,
      "height": 20,
      "lineAppendWidth": 0,
      "lineWidth": 1,
      "matrix": null,
      "name": "test",
      "opacity": 1,
      "strokeOpacity": 1,
      "type": "__GUI_ICON__",
      "visible": true,
      "width": 20,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
    "cfg": Object {
      "animable": true,
      "animating": false,
      "attrs": Object {
        "animable": true,
        "animating": false,
        "capture": true,
        "children": Array [],
        "height": 20,
        "name": "test",
        "type": "__GUI_ICON__",
        "visible": true,
        "width": 20,
        "x": 0,
        "y": 0,
        "zIndex": 0,
      },
      "capture": true,
      "visible": true,
      "zIndex": 0,
    },
    "destroyed": false,
  },
  "isOnlineLink": [Function],
}
`;

exports[`GuiIcon Tests should render correctly icon with https://mdn.alipayobjects.com/huamei_qa8qxu/afts/img/A*P-jqT4U7YrcAAAAAAAAAAAAADmJ7AQ/original.jpg 1`] = `
GuiIcon {
  "_events": Object {},
  "attrs": Object {
    "matrix": null,
    "opacity": 1,
  },
  "cfg": Object {
    "animable": true,
    "animating": false,
    "capture": true,
    "children": Array [],
    "height": 20,
    "name": "test",
    "visible": true,
    "width": 20,
    "x": 0,
    "y": 0,
    "zIndex": 0,
  },
  "destroyed": false,
  "iconImageShape": ImageShape1 {
    "_events": Object {},
    "attrs": Object {
      "animable": true,
      "animating": false,
      "capture": true,
      "children": Array [],
      "fillOpacity": 1,
      "height": 20,
      "lineAppendWidth": 0,
      "lineWidth": 1,
      "matrix": null,
      "name": "test",
      "opacity": 1,
      "strokeOpacity": 1,
      "type": "__GUI_ICON__",
      "visible": true,
      "width": 20,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
    "cfg": Object {
      "animable": true,
      "animating": false,
      "attrs": Object {
        "animable": true,
        "animating": false,
        "capture": true,
        "children": Array [],
        "height": 20,
        "name": "test",
        "type": "__GUI_ICON__",
        "visible": true,
        "width": 20,
        "x": 0,
        "y": 0,
        "zIndex": 0,
      },
      "capture": true,
      "visible": true,
      "zIndex": 0,
    },
    "destroyed": false,
  },
  "isOnlineLink": [Function],
}
`;

exports[`GuiIcon Tests should render correctly icon with https://mdn.alipayobjects.com/huamei_qa8qxu/afts/img/A*f6e6S4OUSdMAAAAAAAAAAAAADmJ7AQ/original.gif 1`] = `
GuiIcon {
  "_events": Object {},
  "attrs": Object {
    "matrix": null,
    "opacity": 1,
  },
  "cfg": Object {
    "animable": true,
    "animating": false,
    "capture": true,
    "children": Array [],
    "height": 20,
    "name": "test",
    "visible": true,
    "width": 20,
    "x": 0,
    "y": 0,
    "zIndex": 0,
  },
  "destroyed": false,
  "iconImageShape": ImageShape1 {
    "_events": Object {},
    "attrs": Object {
      "animable": true,
      "animating": false,
      "capture": true,
      "children": Array [],
      "fillOpacity": 1,
      "height": 20,
      "lineAppendWidth": 0,
      "lineWidth": 1,
      "matrix": null,
      "name": "test",
      "opacity": 1,
      "strokeOpacity": 1,
      "type": "__GUI_ICON__",
      "visible": true,
      "width": 20,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
    "cfg": Object {
      "animable": true,
      "animating": false,
      "attrs": Object {
        "animable": true,
        "animating": false,
        "capture": true,
        "children": Array [],
        "height": 20,
        "name": "test",
        "type": "__GUI_ICON__",
        "visible": true,
        "width": 20,
        "x": 0,
        "y": 0,
        "zIndex": 0,
      },
      "capture": true,
      "visible": true,
      "zIndex": 0,
    },
    "destroyed": false,
  },
  "isOnlineLink": [Function],
}
`;

exports[`GuiIcon Tests should support cross origin for online url 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAZAAAAGQCAYAAACAvzbMAAAOEUlEQVR4Xu3VsQ0AAAjDMPr/0/yQ2exdLKTsHAECBAgQCAILGxMCBAgQIHAC4gkIECBAIAkISGIzIkCAAAEB8QMECBAgkAQEJLEZESBAgICA+AECBAgQSAICktiMCBAgQEBA/AABAgQIJAEBSWxGBAgQICAgfoAAAQIEkoCAJDYjAgQIEBAQP0CAAAECSUBAEpsRAQIECAiIHyBAgACBJCAgic2IAAECBATEDxAgQIBAEhCQxGZEgAABAgLiBwgQIEAgCQhIYjMiQIAAAQHxAwQIECCQBAQksRkRIECAgID4AQIECBBIAgKS2IwIECBAQED8AAECBAgkAQFJbEYECBAgICB+gAABAgSSgIAkNiMCBAgQEBA/QIAAAQJJQEASmxEBAgQICIgfIECAAIEkICCJzYgAAQIEBMQPECBAgEASEJDEZkSAAAECAuIHCBAgQCAJCEhiMyJAgAABAfEDBAgQIJAEBCSxGREgQICAgPgBAgQIEEgCApLYjAgQIEBAQPwAAQIECCQBAUlsRgQIECAgIH6AAAECBJKAgCQ2IwIECBAQED9AgAABAklAQBKbEQECBAgIiB8gQIAAgSQgIInNiAABAgQExA8QIECAQBIQkMRmRIAAAQIC4gcIECBAIAkISGIzIkCAAAEB8QMECBAgkAQEJLEZESBAgICA+AECBAgQSAICktiMCBAgQEBA/AABAgQIJAEBSWxGBAgQICAgfoAAAQIEkoCAJDYjAgQIEBAQP0CAAAECSUBAEpsRAQIECAiIHyBAgACBJCAgic2IAAECBATEDxAgQIBAEhCQxGZEgAABAgLiBwgQIEAgCQhIYjMiQIAAAQHxAwQIECCQBAQksRkRIECAgID4AQIECBBIAgKS2IwIECBAQED8AAECBAgkAQFJbEYECBAgICB+gAABAgSSgIAkNiMCBAgQEBA/QIAAAQJJQEASmxEBAgQICIgfIECAAIEkICCJzYgAAQIEBMQPECBAgEASEJDEZkSAAAECAuIHCBAgQCAJCEhiMyJAgAABAfEDBAgQIJAEBCSxGREgQICAgPgBAgQIEEgCApLYjAgQIEBAQPwAAQIECCQBAUlsRgQIECAgIH6AAAECBJKAgCQ2IwIECBAQED9AgAABAklAQBKbEQECBAgIiB8gQIAAgSQgIInNiAABAgQExA8QIECAQBIQkMRmRIAAAQIC4gcIECBAIAkISGIzIkCAAAEB8QMECBAgkAQEJLEZESBAgICA+AECBAgQSAICktiMCBAgQEBA/AABAgQIJAEBSWxGBAgQICAgfoAAAQIEkoCAJDYjAgQIEBAQP0CAAAECSUBAEpsRAQIECAiIHyBAgACBJCAgic2IAAECBATEDxAgQIBAEhCQxGZEgAABAgLiBwgQIEAgCQhIYjMiQIAAAQHxAwQIECCQBAQksRkRIECAgID4AQIECBBIAgKS2IwIECBAQED8AAECBAgkAQFJbEYECBAgICB+gAABAgSSgIAkNiMCBAgQEBA/QIAAAQJJQEASmxEBAgQICIgfIECAAIEkICCJzYgAAQIEBMQPECBAgEASEJDEZkSAAAECAuIHCBAgQCAJCEhiMyJAgAABAfEDBAgQIJAEBCSxGREgQICAgPgBAgQIEEgCApLYjAgQIEBAQPwAAQIECCQBAUlsRgQIECAgIH6AAAECBJKAgCQ2IwIECBAQED9AgAABAklAQBKbEQECBAgIiB8gQIAAgSQgIInNiAABAgQExA8QIECAQBIQkMRmRIAAAQIC4gcIECBAIAkISGIzIkCAAAEB8QMECBAgkAQEJLEZESBAgICA+AECBAgQSAICktiMCBAgQEBA/AABAgQIJAEBSWxGBAgQICAgfoAAAQIEkoCAJDYjAgQIEBAQP0CAAAECSUBAEpsRAQIECAiIHyBAgACBJCAgic2IAAECBATEDxAgQIBAEhCQxGZEgAABAgLiBwgQIEAgCQhIYjMiQIAAAQHxAwQIECCQBAQksRkRIECAgID4AQIECBBIAgKS2IwIECBAQED8AAECBAgkAQFJbEYECBAgICB+gAABAgSSgIAkNiMCBAgQEBA/QIAAAQJJQEASmxEBAgQICIgfIECAAIEkICCJzYgAAQIEBMQPECBAgEASEJDEZkSAAAECAuIHCBAgQCAJCEhiMyJAgAABAfEDBAgQIJAEBCSxGREgQICAgPgBAgQIEEgCApLYjAgQIEBAQPwAAQIECCQBAUlsRgQIECAgIH6AAAECBJKAgCQ2IwIECBAQED9AgAABAklAQBKbEQECBAgIiB8gQIAAgSQgIInNiAABAgQExA8QIECAQBIQkMRmRIAAAQIC4gcIECBAIAkISGIzIkCAAAEB8QMECBAgkAQEJLEZESBAgICA+AECBAgQSAICktiMCBAgQEBA/AABAgQIJAEBSWxGBAgQICAgfoAAAQIEkoCAJDYjAgQIEBAQP0CAAAECSUBAEpsRAQIECAiIHyBAgACBJCAgic2IAAECBATEDxAgQIBAEhCQxGZEgAABAgLiBwgQIEAgCQhIYjMiQIAAAQHxAwQIECCQBAQksRkRIECAgID4AQIECBBIAgKS2IwIECBAQED8AAECBAgkAQFJbEYECBAgICB+gAABAgSSgIAkNiMCBAgQEBA/QIAAAQJJQEASmxEBAgQICIgfIECAAIEkICCJzYgAAQIEBMQPECBAgEASEJDEZkSAAAECAuIHCBAgQCAJCEhiMyJAgAABAfEDBAgQIJAEBCSxGREgQICAgPgBAgQIEEgCApLYjAgQIEBAQPwAAQIECCQBAUlsRgQIECAgIH6AAAECBJKAgCQ2IwIECBAQED9AgAABAklAQBKbEQECBAgIiB8gQIAAgSQgIInNiAABAgQExA8QIECAQBIQkMRmRIAAAQIC4gcIECBAIAkISGIzIkCAAAEB8QMECBAgkAQEJLEZESBAgICA+AECBAgQSAICktiMCBAgQEBA/AABAgQIJAEBSWxGBAgQICAgfoAAAQIEkoCAJDYjAgQIEBAQP0CAAAECSUBAEpsRAQIECAiIHyBAgACBJCAgic2IAAECBATEDxAgQIBAEhCQxGZEgAABAgLiBwgQIEAgCQhIYjMiQIAAAQHxAwQIECCQBAQksRkRIECAgID4AQIECBBIAgKS2IwIECBAQED8AAECBAgkAQFJbEYECBAgICB+gAABAgSSgIAkNiMCBAgQEBA/QIAAAQJJQEASmxEBAgQICIgfIECAAIEkICCJzYgAAQIEBMQPECBAgEASEJDEZkSAAAECAuIHCBAgQCAJCEhiMyJAgAABAfEDBAgQIJAEBCSxGREgQICAgPgBAgQIEEgCApLYjAgQIEBAQPwAAQIECCQBAUlsRgQIECAgIH6AAAECBJKAgCQ2IwIECBAQED9AgAABAklAQBKbEQECBAgIiB8gQIAAgSQgIInNiAABAgQExA8QIECAQBIQkMRmRIAAAQIC4gcIECBAIAkISGIzIkCAAAEB8QMECBAgkAQEJLEZESBAgICA+AECBAgQSAICktiMCBAgQEBA/AABAgQIJAEBSWxGBAgQICAgfoAAAQIEkoCAJDYjAgQIEBAQP0CAAAECSUBAEpsRAQIECAiIHyBAgACBJCAgic2IAAECBATEDxAgQIBAEhCQxGZEgAABAgLiBwgQIEAgCQhIYjMiQIAAAQHxAwQIECCQBAQksRkRIECAgID4AQIECBBIAgKS2IwIECBAQED8AAECBAgkAQFJbEYECBAgICB+gAABAgSSgIAkNiMCBAgQEBA/QIAAAQJJQEASmxEBAgQICIgfIECAAIEkICCJzYgAAQIEBMQPECBAgEASEJDEZkSAAAECAuIHCBAgQCAJCEhiMyJAgAABAfEDBAgQIJAEBCSxGREgQICAgPgBAgQIEEgCApLYjAgQIEBAQPwAAQIECCQBAUlsRgQIECAgIH6AAAECBJKAgCQ2IwIECBAQED9AgAABAklAQBKbEQECBAgIiB8gQIAAgSQgIInNiAABAgQExA8QIECAQBIQkMRmRIAAAQIC4gcIECBAIAkISGIzIkCAAAEB8QMECBAgkAQEJLEZESBAgICA+AECBAgQSAICktiMCBAgQEBA/AABAgQIJAEBSWxGBAgQICAgfoAAAQIEkoCAJDYjAgQIEBAQP0CAAAECSUBAEpsRAQIECAiIHyBAgACBJCAgic2IAAECBATEDxAgQIBAEhCQxGZEgAABAgLiBwgQIEAgCQhIYjMiQIAAAQHxAwQIECCQBAQksRkRIECAgID4AQIECBBIAgKS2IwIECBAQED8AAECBAgkAQFJbEYECBAgICB+gAABAgSSgIAkNiMCBAgQEBA/QIAAAQJJQEASmxEBAgQICIgfIECAAIEkICCJzYgAAQIEBMQPECBAgEASEJDEZkSAAAECAuIHCBAgQCAJCEhiMyJAgAABAfEDBAgQIJAEBCSxGREgQICAgPgBAgQIEEgCApLYjAgQIEBAQPwAAQIECCQBAUlsRgQIECAgIH6AAAECBJKAgCQ2IwIECBAQED9AgAABAklAQBKbEQECBAgIiB8gQIAAgSQgIInNiAABAgQExA8QIECAQBIQkMRmRIAAAQIC4gcIECBAIAkISGIzIkCAAAEB8QMECBAgkAQEJLEZESBAgICA+AECBAgQSAICktiMCBAgQEBA/AABAgQIJAEBSWxGBAgQICAgfoAAAQIEkoCAJDYjAgQIEBAQP0CAAAECSUBAEpsRAQIECAiIHyBAgACBJCAgic2IAAECBB50NwGRNp50KwAAAABJRU5ErkJggg=="`;
